recipes/repo.rb

```diff
@@ -1,49 +1,186 @@
 # Update packages before starting cookbook run.
 apt_update
 
-package "reprepro"
+package 'reprepro'
 
-package "openssh-server"
+package 'openssh-server'
 
-host_keys = data_bag_item("ros_buildfarm_host_keys", "repo")[node.chef_environment]
+host_keys = data_bag_item('ros_buildfarm_host_keys', 'repo')[node.chef_environment]
 %w(dsa ecdsa ed25519 rsa).each do |type|
   file "/etc/ssh/ssh_host_#{type}_key" do
-    content host_keys[type]["private"]
-    mode "0600"
+    content host_keys[type]['private']
+    mode '0600'
   end
   file "/etc/ssh/ssh_host_#{type}_key.pub" do
-    content host_keys[type]["public"]
-    mode "0644"
+    content host_keys[type]['public']
+    mode '0644'
   end
 end
 
 # Update attributes to get a "building repository" agent instead of a generic
 # "buildagent".
-node.default['ros_buildfarm']['agent']['nodename'] = "building_repository"
+node.default['ros_buildfarm']['agent']['nodename'] = 'building_repository'
 node.default['ros_buildfarm']['agent']['executors'] = 1
-node.default['ros_buildfarm']['agent']['labels'] = ["building_repository"]
-include_recipe "ros_buildfarm::agent"
+node.default['ros_buildfarm']['agent']['labels'] = ['building_repository']
+include_recipe 'ros_buildfarm::agent'
 
-agent_username = node["ros_buildfarm"]["agent"]["agent_username"]
+agent_username = node['ros_buildfarm']['agent']['agent_username']
 
 # Create web root and web directories
-# TODO: why are these two 644 and the next two 755?
-# For now culting it from the puppet config.
 %w(/var/repos /var/repos/ubuntu).each do |dir|
   directory dir do
     owner agent_username
     group agent_username
-    mode "0644"
+    mode '0755'
   end
 end
 
 %w(docs rosdistro_cache status_page).each do |dir|
   directory "/var/repos/#{dir}" do
     owner agent_username
     group agent_username
-    mode "0755"
+    mode '0755'
   end
 end
 
-package "python3-yaml"
-package "python3-debian"
+package 'python3-yaml'
+package 'python3-debian'
+
+# upload scripts and keys
+directory "/home/#{agent_username}/upload_triggers" do
+  owner agent_username
+  group agent_username
+end
+# TODO: at some point this file will need to be templatized and only installed if needed.
+cookbook_file "/home/#{agent_username}/upload_repo.bash" do
+  source 'upload_repo.bash'
+  owner agent_username
+  group agent_username
+  mode '0700'
+end
+data_bag('ros_buildfarm_upload_keys').each do |id|
+  key = data_bag_item('ros_buildfarm_upload_keys', id)
+  file "/home/#{agent_username}/upload_triggers/#{key['name']}" do
+    content key['content']
+    owner agent_username
+    group agent_username
+    mode '0600'
+  end
+end
+
+# Configure GPG for reprepro
+# .gnupg/gpg.conf
+directory "/home/#{agent_username}/.gnupg" do
+  owner agent_username
+  group agent_username
```

Comment:
Do you think it's important that the GPG *config* is not readable by other users?

Reply:
Same as the .gpg directory I think we'd want this to be `0700` no?

Reply:
`0700` is executable bit no? In directories is being used to access items inside even if you can not list items inside, I think. Do we want that instead of complete blocking access to non user?

Reply:
Executable bit on directories is required for the directory to be traversable. That is, in order to access foo/bar/baz both foo and bar need to be traversable.